### DISCUSSION TITLE
Track cells with predefined segmentation masks

### DISCUSSION THREAD

**Discussion by tuanpham96** [SchmollerLab/Cell_ACDC#181] [Q&A]
I have data that have already been segmented for each day but not aligned across days for either mask ids nor field of views. I was wondering whether/how I can reshape my data somehow to load into Cell_ACDC for aligning the masks. 

Plus, if the tracking is automated, is it possible to have manual correction afterwards?

Thanks!

**Comment by ElpadoCan** [comment]
Hi @tuanpham96, you can absolutely do the tracking on previously segmented data. And you can choose to do it either automatically or manually, or first automatically and then manually correct.

You are going to need to structure your data in a specific way. If you have the microscopy files I recommend using the module
 `0. Create data structure from microscopy file(s)...` of Cell-ACDC (first button on the little launcher that opens when you launch Cell-ACDC). See the section called `Create required data structure from microscopy file(s)` of the [user manual](https://github.com/SchmollerLab/Cell_ACDC/blob/main/UserManual/Cell-ACDC_User_Manual.pdf).

Alternatively, you can create the data structure manually. Have a look at the section called `Manually create data structure from microscopy files(s)` in our user manual [here](https://github.com/SchmollerLab/Cell_ACDC/blob/main/UserManual/Cell-ACDC_User_Manual.pdf).

Briefly, you need one .tif file per position per channel (if you have multiple fluorescent channels for example). Then you put the .tif file(s) into a subfolder with relative path `...\Position_1\Images`. 

Once you have the structure, inside the `Images` folder, you also put the segmentation file that needs to be a NumPy array in `.npz` format. If you have other formats we have a utility for converting it to `.npz`. You can find conversion utilities from the menu on the top menu bar of the Cell-ACDC launcher called `Utilities --> Convert ...`. 

The segmentation file needs to have the same start name (which I call the basename) of the .tif file and it needs to include the word `segm` after the common basename. So for example let's suppose you have a .tif file called `my_experiment_phase_contrast.tif`, then valid segmentation file names are for example `my_experiment_segm.npz`, or  `my_experiment_segm_phase_contrast.npz`, or whatever text you want to have after `segm_`. 

Finally, if you want to work with just the segmentation file (without loading any signal like phase contrast, bright field, or fluorescent), I just thought now that it could be a cool feature to add. Essentially, I could add a feature that allows you to simply load a segmentation file from any folder you like, with any format you like. Cell-ACDC would then internally convert it to `npz` and it would work on that file. 

I hope it's clear enough and what you think about loading only the segmentation file.

Cheers,
Francesco

**Reply by tuanpham96** [reply]
Hi Francesco,

Thank you for your prompt answers! 

I hope to verify a few terminologies and data structure. 

For your info, I work with calcium imaging data across days, each day segmented with [`suite2p`](https://github.com/MouseLand/suite2p/). For each day, the output segmentation is a collection of cells' binary masks.

**Q1**: Is your use of `Position` and my use of `Day` equivalent?

 **Q2**: Say I have 1 maximum (or mean) projected image with only 1 channel each day. Is the following the correct pseudo directory structure?

```
.
├── Day0
│   └── Images
│       ├── day0_maxproj.tif
│       └── day0_maxproj_segm.npz
├── Day1
│   └── Images
│       ├── day1_maxproj.tif
│       └── day1_maxproj_segm.npz
└── Day2
    └── Images
        ├── day2_maxproj.tif
        └── day2_maxproj_segm.npz 
```

**Q3**: By only `_segm.npz`, and not followed by other suffices (like `_phase_contrast` or `_fluorescent`):
- **Q3a**: Is that a collection of binary masks? If it's more involved, do you have a sample file that I can model after?
  - Is it just a collection of `numpy.array[int64]` objects (one per ROI) of size `height x width` with only `0` and `1`?
  - Or, is it a collection of lists of pixels for each ROI? 
- **Q3b**: For `suite2p` segmentation, there are instances of cells whose masks overlap slightly, usually at the edges. It's possible to remove these overlapping pixels to have a clean binary mask, but doing such might distort the shape of those cells' masks. So my question is, does this affect the program?

Thanks!

**Reply by ElpadoCan** [reply]
Hi @tuanpham96, let me answer point by point

> Q1: Is your use of Position and my use of Day equivalent?
> Q2: Say I have 1 maximum (or mean) projected image with only 1 channel each day. Is the following the correct pseudo directory structure?

You could do that, however, there is no tracking between different Positions, so it would not work with what you want to achieve (I'm guessing you want track over time, i.e. over days?). Also, you cannot call the Position folder `Day0` you would need to call it `Position_1`. The idea of positions is that we often acquire multiple single cells growing over time into a colony, and each single cell is one Position. In Fiji they are called Series for example. The idea of the `Images` folder instead is that inside the Position folder I often have other type of downstream analysis (of the same Position), each in its separate folder. If you want to perform tracking, then each .tif file needs to have the days concatenated, so if you have 5 days of (512,512) then the data shape saved in the .tif file would be (5,512,512). Note that you could also have 4D data in Cell-ACDC, which is z-stacks over time. For 5 days with 20 z-slices you would have a .tif file containing data with shape (5,20,512,512). 

> Q3: By only _segm.npz, and not followed by other suffices (like _phase_contrast or _fluorescent):

You are allowed to have other suffices which can be any text. This allows having multiple different segmentation files. If you compute the metrics with Cell-ACDC (mean, max, area etc.) there will be an additional file ending with the same suffices and preceded by `acdc_output`, for example, the file ending with `segm_phase_contrast.npz` will generate the metrics file `acdc_output_phase_contrast.csv`, which is a table in CSV format. 

> Q3a: Is that a collection of binary masks? If it's more involved, do you have a sample file that I can model after?
Is it just a collection of numpy.array[int64] objects (one per ROI) of size height x width with only 0 and 1?
Or, is it a collection of lists of pixels for each ROI?

The `segm.npz` file is an array of integers (usually called labels, IDs, or Cell_ID) with the same exact shape as the image you segmented. In this array, each pixel belonging to the background will have value of 0, while pixels belonging to the same segmented objects (also called ROIs in Fiji) will have a specific ID (an integer). This is the typical output of instance segmentation algorithms such as cellpose, yeaz, stardist etc. If you need the ROIs files (compatible with Fiji) we have an utility for the conversion of segm.npz to ROIs. We also have utilities to convert to .tif for example (which is also compatible with Fiji or cellpose). I can send you an example Position folder if you give me a google drive email, or you can write me an email to padovaf@tcd.ie and I can reply with a position folder. You should be able to download an example dataset also directly from the Welcome Guide of Cell-ACDC. Alternatively, the datasets we used for our publication are freely accessible and public on zenodo at this link https://zenodo.org/record/6795124

You could download the `test_data_budding_yeast` dataset. In that dataset an example of data structure is `test_data_budding_yeast/TimeLapse_2D/MIA_KC_htb1_mCitrine_flu_control_labeled/Position_1/Images` where the segmentation file is called `KC_2021-03-19-ASY020-1_00nM_SCGE-Subset_s1_segm.npz`.

If you want to load a `segm.npz` file as an example in Python you can use this simple script (it will print the shape and the data type of the array):
```
import numpy as np

segm_data = np.load('path_to_segm.npz')['arr_0']

print(segm_data.shape, segm_data.dtype)
```

Anyway, I can easily implement a utility that would load your data, reshape it and save it into the proper data structure required for Cell-ACDC. I just need to know from which data structure you want to start. I recommend starting from the raw microscopy file, but we would also start from the output of suite2p. Maybe you share with me at my email padovaf@tcd.ie (Google Drive is fine, but you can use whatever cloud you prefer) a sample of the data that you would like to load into Cell-ACDC? Let me know how you want to proceed, thanks! 

Cheers,
Francesco

**Reply by ElpadoCan** [reply]
I forgot to answer one question:

> Q3b: For suite2p segmentation, there are instances of cells whose masks overlap slightly, usually at the edges. It's possible to remove these overlapping pixels to have a clean binary mask, but doing such might distort the shape of those cells' masks. So my question is, does this affect the program?

I would have to look into the output of suite2p, I am not familiar with this library. If you could send me sample data I can run some tests. In principle, if you need to have objects that do not touch each other there you could for example slightly shrink all the masks by some constant (the operation is called `erosion` and it's a pretty common operation in morphological image processing) until they don't touch anymore. I'm not familiar with the type of analysis you need to do, but usually, it is not a problem if masks touch each other.  In any case in Cell-ACDC, there is a function to shrink or expand objects, it is performed on specific objects but I can easily implement some keyboard modifier to apply the operation to all the masks at once.

**Reply by tuanpham96** [reply]
Thank you so much for your detailed instructions and pointing me to the demo data! I was able to follow and load the data into the program, but I haven't played around much yet. 

I'll email you the data later this week. The dataset would have 2 planes of 512 x 512 across 5 days, and with this one there doesn't seem to be any overlap. Hence both the shape of the tiff and npz files are `5 x 2 x 512 x 512`. 

In the meantime, I have a few of follow-up questions:

**Q1**: for tracking, does the program take into account the z-dim? For this dataset, the z-distance is actually very far to sample the space, so we'd only hope to track across days. 

**Q2**: for tracking, does the program do an initial step of field-of-view alignment? I saw that there's mentioning of some `align` option for the `Data prep` step but I'm not sure whether such action can be applied in this stage. 

**Q3**: does the tracking base on previous frame as reference? or is it possible to base on a chosen frame? The reason is because sometimes there's a bad recording day. 

**Q4**: I was trying to see whether I can have multiple frames side by side and optionally showing only a shadow (i.e. only faint contour or an opaque version) on top of one another. For the limited time I played around with the program, it seems that the 2 frames contours/IDs overlap on each other a bit overwhelmingly and hard to distinguish between the ROIs from the 2 planes. I'll play around with it more to discover how to do tracking properly. 

**Q5**: from your definition of `Position_X`, is it right to say that the positions are independent of each other from the computational standpoint?

**Reply by ElpadoCan** [reply]
Hi @tuanpham96, sorry for the late reply, but for some reason, I didn't get the notification of your comment and I saw it only now.

Anyway, let me answer your questions:

> I'll email you the data later this week. The dataset would have 2 planes of 512 x 512 across 5 days, and with this one there doesn't seem to be any overlap. Hence both the shape of the tiff and npz files are 5 x 2 x 512 x 512.

Yesterday I released `v1.2.4rc49` where you can create the required data structure starting from .tif files where each .tif file is one time-point, in your case one file per day. However, I didn't understand if you are planning to open only the segmentation file or also the signal. I would recommend loading both. Once I have your data I can see what can be done.

> Q1: for tracking, does the program take into account the z-dim? For this dataset, the z-distance is actually very far to sample the space, so we'd only hope to track across days.

Cell-ACDC has three trackers as of now: YeaZ, Cell-ACDC and [trackpy](https://soft-matter.github.io/trackpy/v0.5.0/). Both YeaZ and Cell-ACDC do not work with 3D z-stack objects, but they can be adapted if needed. Cell-ACDC tracker was developed more for real-time application because it is fast and it can point out tracking mistakes. Trackpy on the other hand works only with the distances between objects (and eventually the direction of the movement between consecutive frames) so it also works in 3D, but I never tried it before. In 2D over time it works pretty well.  
If possible for your application I would start by segmenting 2D objects (you can do that by keeping your data 4D and selecting which z-slice you want to segment in the data prep module). Note that you cannot use trackpy in real-time, so you first track everything, then open the GUI, disable real-time tracking and you can just visualize the result of the tracker.

> Q2: for tracking, does the program do an initial step of field-of-view alignment? I saw that there's mentioning of some align option for the Data prep step but I'm not sure whether such action can be applied in this stage.

The alignment that you can apply with Cell-ACDC is a XY frame-to-frame alignment based on cross-correlation (we use the function [`skimage.registration.phase_cross_correlation`](https://scikit-image.org/docs/stable/api/skimage.registration.html#skimage.registration.phase_cross_correlation). This alignment shifts the field-of-view in X and Y by a specific amount. Whether it works well, or it makes tracking better needs to be tested. For our datasets, it works pretty well, but I have seen cases in which a sudden flash of light or weird objects passing by are causing the alignment to completely go wild.

> Q3: does the tracking base on previous frame as reference? or is it possible to base on a chosen frame? The reason is because sometimes there's a bad recording day.

As of now, all the tree trackers work on frame-to-frame. We are working on developing a tracker that can take more frames into account but it's gonna take some time. In this case, I would completely remove that day from the data. Would that be an option for you?

> Q4: I was trying to see whether I can have multiple frames side by side and optionally showing only a shadow (i.e. only faint contour or an opaque version) on top of one another. For the limited time I played around with the program, it seems that the 2 frames contours/IDs overlap on each other a bit overwhelmingly and hard to distinguish between the ROIs from the 2 planes. I'll play around with it more to discover how to do tracking properly.

I cannot visualize the problem you are having so maybe you can attach a screenshot? For the tracking in your case, I would first try trackpy from module `3. Launch GUI...`. You load the data, you switch to "Segmentation and tracking" mode from the top toolbar and then you go to the menu on the top menu bar called `Tracking --> Repeat tracking on multiple frames...`. Then you can choose how many frames to track and which tracker to use. If the objects do not move much between frames and there is overlap then both YeaZ and Cell-ACDC should work fine, but I think trackpy will be better. 

I have a feeling that the 2 planes you are talking about would be two different set of objects and not z-slices of the same object? Sorry but I am not familiar with calcium imaging, but it would be very cool to add this application to Cell-ACDC. 

> Q5: from your definition of Position_X, is it right to say that the positions are independent of each other from the computational standpoint?

Exactly, Positions are completely independent.

I hope it's clear enough and let me know how your progress goes, thanks! 

Cheers,
Francesco

**Reply by tuanpham96** [reply]
Thanks for your answers!

> However, I didn't understand if you are planning to open only the segmentation file or also the signal. I would recommend loading both. Once I have your data I can see what can be done.

I emailed you the data with a GDrive link, and along with how I converted the data to prepare for Cell-ACDC. Let me know if you have issues opening it. You can also refer to the [`suite2p` output documentation](https://suite2p.readthedocs.io/en/latest/outputs.html) if you have question about the format. We can communicate via email about the particulars of the dataset or set up a video chat to iron out the details. 

For each day, the mean-projected images aren't great, this is because there will be a subset of cells that usually have a lot of activity throughout the experiments, but some could be only sparsely activated but still important for analysis later on. The max-projected images are usually cropped so I don't include them. 

I included the mean image as the `.tif` file and the segmented file as the the `.npz` files. 

---

> Cell-ACDC has three trackers as of now: YeaZ, Cell-ACDC and trackpy. [...]

> If the objects do not move much between frames and there is overlap then both YeaZ and Cell-ACDC should work fine, but I think trackpy will be better.

Thanks for the descriptions & suggestions. I don't work with microbiology data so I'd have to take some time familiarize with them and trying them out to see whether they would work with calcium data. 

> If possible for your application I would start by segmenting 2D objects (you can do that by keeping your data 4D and selecting which z-slice you want to segment in the data prep module). 

I included 2 options actually in regards to this in the code I sent you:

- The first option was `plane_as_position=False`, in which the plane is included in 4D array. E.g `Position_1/both_planes.tiff` has dim `(5 days, 2 planes, 512, 512)`
- The second option was `plane_as_position=True` in order to keep the 4D matrix but separate each plane into a separate `Position_X` folder. E.g. `Position_1/plane0.tif` has dim `(5 days, 1 plane, 512, 512)` and `Position_2/plane1.tif` has dim `(5 days, 1 plane, 512, 512)`

**Q1**: By the way, does it matter whether the starting index of `Position_X` is `1` or `0`?

---

> I have a feeling that the 2 planes you are talking about would be two different set of objects and not z-slices of the same object? 

Short answer: yes, longer answer below. 

> Exactly, Positions are completely independent.

Based on these points, I would assume `plane_as_position=True` is the correct way to do it.

---

> The alignment that you can apply with Cell-ACDC is a XY frame-to-frame alignment based on cross-correlation (we use the function skimage.registration.phase_cross_correlation. This alignment shifts the field-of-view in X and Y by a specific amount. Whether it works well, or it makes tracking better needs to be tested. For our datasets, it works pretty well, but I have seen cases in which a sudden flash of light or weird objects passing by are causing the alignment to completely go wild.

**Q2**: Do you recommend phase-corrected before inputs into `acdc`?

---

> As of now, all the tree trackers work on frame-to-frame. We are working on developing a tracker that can take more frames into account but it's gonna take some time. In this case, I would completely remove that day from the data. Would that be an option for you?

Possibly, though this might depend on the experiment and dataset. The dataset I sent you contains 2 animals: `ME04` and `ME05`. The `ME05_20220604`  should be the bad one because of something blocking the field of view of that day but the following day was fine. For now, I think we can ignore that day of that animal but I included it as an example if you want to take a look.
 
---

> I have a feeling that the 2 planes you are talking about would be two different set of objects and not z-slices of the same object? Sorry but I am not familiar with calcium imaging, but it would be very cool to add this application to Cell-ACDC.

The experimenter wants to sample that brain area to have as many neurons as possible to record population activity in a session, (from my basic understanding) the objective can move its focus in the z-direction to achieve this. And the z-planes are sampled to be far from each other so as not to have activity contamination between the planes, but not too far in order to not sacrifice too much of the sampling rate in that session. 

So to answer you question, technically it's z-slices of the same "object". But from a computational standpoint, they can be considered different "object".

**Q3**: Just out of curiosity, do the data that `acdc` is originally meant for usually have z-planes of the same object?

**Reply by ElpadoCan** [reply]
Hi @tuanpham96, I got your data and I will do some tests in the next few days. I'm at a conference for two days so I will not have much time, but when I'm back I will definitely check it out. 

I'm very excited to expand Cell-ACDC scope to calcium imaging so rest assured that we will make it work and then we see if it's useful for you. Here are my answers:

> Q1: By the way, does it matter whether the starting index of Position_X is 1 or 0?

It doesn't matter, but I always started at 1. 

> Q2: Do you recommend phase-corrected before inputs into acdc?

I would definitely try it. If it doesn't make things worse, it will probably be beneficial.

> Q3: Just out of curiosity, do the data that acdc is originally meant for usually have z-planes of the same object?

Cell-ACDC can work with 2D, 3D (either single z-stack snapshot one per position, or 2D data over time) or 4D data (z-stack over time). A z-stack is a sectioning of the same sample that usually takes a few minutes to be taken with a microscope. 

I hope I answered everything and I will keep you updated with the tests I will do with your data. Have a good week! 

Cheers,
Francesco

**Reply by ElpadoCan** [reply]
Hi @tuanpham96,

I had a look at your data and I didn't know you already tried to get the right structure for Cell-ACDC, nice job! 

So the `plane_inside` data can be loaded without problem. The `plane_as_position` instead has the wrong shape. The shape of the .tif file that you created is `(5,1,512,512)`, while the correct shape is `(5,512,512)` without the 1 for the plane. This can easily be achieved by calling the NumPy function [`numpy.squeeze`](https://numpy.org/doc/stable/reference/generated/numpy.squeeze.html) which automatically removes axes with dimension 1.

Basically, you can add the squeeze function at line 151 of your `utils.py` file as follows:

```python
imwrite(file_name, np.squeeze(mat))
```

The `_segm.npz` file instead is correct in both cases. Just to be sure I would also add the squeeze function at line 144 of your `utils.py` file as follow:

```python
np.savez(file_name, np.squeeze(mat))
```

I hope it's clear and let me know if it works! 

By the way, since I am guessing you probably are going to need some new features, I think the easiest would be for you to install from source instead of from PyPi. This way you can get new features before they are released and we can release them once you tested them.

To install from source you need to clone the repo with `git` and then install the locally cloned repo. You could do this in a separate environment (that you can call for example `acdc-source` or whatever you like). This way you have both the release version and the developer version.

You can find instructions on how to install from source [here](https://github.com/SchmollerLab/Cell_ACDC#install-from-source). Let me know if you need help with that.

Cheers,
Francesco

**Reply by ElpadoCan** [reply]
I forgot one thing. At the moment you are working with only one .tif file (one channel) called for example `ME04_plane0.tif` where `ME04_plane0` is the basename and there is no channel name. This is of course fine for one channel, but remember that you can add a channel name after the basename, for example, `ME04_plane0_GFP.tif`. This is handy and required with two or more channels where you could have another .tif file called `ME04_plane0_mNeon.tif` or whatever the channel name is. The channel name can be any text you want, I usually use the fluorophore name, but you could for example call it `nuclei`. 

For the segmentation file, it is the same idea. You can have `ME04_plane0_segm.npz` or you could also have `ME04_plane0_segm_GFP.npz` if you plan to have more than one segmentation file (for example a segmentation mask of another channel, or one 2D segmentation file and one 3D). 

I hope it's clear enough and let me know if you have questions.

Cheers,
Francesco

**Reply by tuanpham96** [reply]
> The shape of the .tif file that you created is (5,1,512,512), while the correct shape is (5,512,512) without the 1 for the plane.
 
I don't remember why, but I think it didn't pick up that there's 1 plane and 5 frames when I was doing `(5,512,512)` in the first place, hence I didn't `squeeze` it. But it was probably when I was tinkering around. Anyway, I can add `squeeze` back. 

> I think the easiest would be for you to install from source instead of from PyPi.

Good point. I'm currently using it from `venv` anyway so it's easy enough to switch. 

> This is handy and required with two or more channels

Thanks for the tip! This would be very helpful as there are others in the lab doing 2 channels. 

> if you plan to have more than one segmentation file (for example a segmentation mask of another channel, or one 2D segmentation file and one 3D).

So generally we'd have 1 segmentation file in cases of 2 channels. How does the program handle 2 tifs and 1 segmentation file? Does it load everything in to track?

**Reply by ElpadoCan** [reply]
Hi @tuanpham96,

The squeeze should work because to load your data I also added squeeze to Cell-ACDC and I was able to load it. Note that even if the shape is `(5,512,512)` you are still required to provide the number of z-slices, and by inserting 1 z-slice Cell-ACDC will ignore that and expects data with shape `(T, Y, X)`. 

>So generally we'd have 1 segmentation file in cases of 2 channels. How does the program handle 2 tifs and 1 segmentation file? Does it load everything in to track?

You can choose how many channels to load for the computation of numerical features based on the loaded segmentation file. If for example you have two channels `GFP` and `mNeon` and one segmentation file ending with `segm_cells.npz` you can load this segmentation file, then you can load the two channels and compute the metrics. Cell-ACDC will save a file ending with `acdc_output_cells.csv` containing features of both channels (for example `GFP_mean` and `mNeon_mean`) based on the `segm_cells.npz` masks. All of this can be done from the menu on the top menubar of the main GUI (module 3. from the launcher) called `Measurements --> Set measurements...`. This will open a dialogue where you can choose to load channels and choose which features to save. Each feature has an info button with an explanation about that feature.

Let me know how it goes, thanks! 

Cheers,
Francesco
